[excel] (PivotTable) PivotTable object model article (#485)
OfficeDev/office-scripts-docs: docs/develop/pivottables.md

---
title: 'Work with PivotTables in Office Scripts'

> description: 'The object model for PivotTables in the Office Scripts JavaScript API'

Reviewer: ```suggestion
description: 'Learn about the object model for PivotTables in the Office Scripts JavaScript API.'
```
Description metadata should include a CTA, if possible. Also, since it's included in SERP, end with a period.

ms.date: 04/20/2022
ms.localizationpriority: medium
---

# Work with PivotTables in Office Scripts

> PivotTables let you analyze large collections of data faster. With their power comes complexity. The Office Scripts APIs let you customize a PivotTable to suit your needs, but the scope of the API set makes getting started a challenge. This article demonstrates how to perform common PivotTable tasks and explains important classes and methods.

Reviewer: ```suggestion
PivotTables let you analyze large collections of data faster . With their power comes complexity. The Office Scripts APIs let you customize a PivotTable to suit your needs, but the scope of the API set makes getting started a challenge. This article demonstrates how to perform common PivotTable tasks and explains important classes and methods.
```
Consider changing "faster" to "quickly" or "efficiently" or another appropriate adverb. Comparative adjectives are used to compare differences between the attributes in two nouns. I

> [!NOTE]

> > To better understand context for the terms used by the APIs, please read Excel's PivotTable documentation first. Start with [Create a PivotTable to analyze worksheet data](https://support.microsoft.com/office/a9a84538-bfe9-40a9-a8e9-f99134456576).

Reviewer: ```suggestion
> To better understand context for the terms used by the APIs, read Excel's PivotTable documentation first. Start with [Create a PivotTable to analyze worksheet data](https://support.microsoft.com/office/a9a84538-bfe9-40a9-a8e9-f99134456576).
```

The [PivotTable](/javascript/api/office-scripts/excelscript/excelscript.pivottable) is the central object for PivotTables in the Office Scripts API.

- The [Workbook](/javascript/api/office-scripts/excelscript/excelscript.workbook) object has a collection of all the [PivotTables](/javascript/api/office-scripts/excelscript/excelscript.pivottable). Each [Worksheet](/javascript/api/office-scripts/excelscript/excelscript.worksheet) also contains a PivotTable collection that's local to that sheet.
- A [PivotTable](/javascript/api/office-scripts/excelscript/excelscript.pivottable) contains [PivotHierarchies](/javascript/api/office-scripts/excelscript/excelscript.pivothierarchy). A hierarchy can be thought of as a column in a table.
- [PivotHierarchies](/javascript/api/office-scripts/excelscript/excelscript.pivothierarchy) can be added as rows or columns ([RowColumnPivotHierarchy](/javascript/api/office-scripts/excelscript/excelscript.rowcolumnpivothierarchy)), data ([DataPivotHierarchy](/javascript/api/office-scripts/excelscript/excelscript.datapivothierarchy)), or filters ([FilterPivotHierarchy](/javascript/api/office-scripts/excelscript/excelscript.filterpivothierarchy)).

> - Each [PivotHierarchy](/javascript/api/office-scripts/excelscript/excelscript.pivothierarchy) contains one and only one [PivotField](/javascript/api/office-scripts/excelscript/excelscript.pivotfield). PivotTable structures outside of Excel may contain multiple fields per hierarchy, so this design exists to support future options. For Office Scripts, fields and hierarchies map to the same information.

Reviewer: ```suggestion
- Each [PivotHierarchy](/javascript/api/office-scripts/excelscript/excelscript.pivothierarchy) contains only one [PivotField](/javascript/api/office-scripts/excelscript/excelscript.pivotfield). PivotTable structures outside of Excel may contain multiple fields per hierarchy, so this design exists to support future options. For Office Scripts, fields and hierarchies map to the same information.
```

- A [PivotField](/javascript/api/office-scripts/excelscript/excelscript.pivotfield) contains multiple [PivotItems](/javascript/api/office-scripts/excelscript/excelscript.pivotitem). Each PivotItem is a unique value in the field. Think of each item as a value in the table column. Items could also be aggregated values, such as sums, if the field is being used for data.
- The [PivotLayout](/javascript/api/office-scripts/excelscript/excelscript.pivotlayout) defines how the [PivotFields](/javascript/api/office-scripts/excelscript/excelscript.pivotfield) and [PivotItems](/javascript/api/office-scripts/excelscript/excelscript.pivotitem) are displayed.
- [PivotFilters](/javascript/api/office-scripts/excelscript/excelscript.pivotfilters) filter data from the [PivotTable](/javascript/api/office-scripts/excelscript/excelscript.pivottable) using different criteria.

> Let's look at how these relationships work in practice. The following data describes fruit sales from various farms. It is the base for all the examples in this article. Use <a href="pivottable-sample.xlsx">pivottable-sample.xlsx</a> to follow along.

Reviewer: ```suggestion
Look at how these relationships work in practice. The following data describes fruit sales from various farms. It's the base for all the examples in this article. Use <a href="pivottable-sample.xlsx">pivottable-sample.xlsx</a> to follow along.
```
Per MSFT Voice & Style, we use active tense, second person.

:::image type="content" source="../images/pivottable-raw-data.png" alt-text="A collection of fruit sales of different types from different farms.":::

## Create a PivotTable with fields

> PivotTables are created with references to existing data. Both ranges and tables can be the source for a PivotTable. They also need a place to exist in the workbook. Since the size of PivotTable is dynamic, only the upper-left corner of the destination range is specified.

Reviewer 2: Missing article 
```suggestion
PivotTables are created with references to existing data. Both ranges and tables can be the source for a PivotTable. They also need a place to exist in the workbook. Since the size of a PivotTable is dynamic, only the upper-left corner of the destination range is specified.
```

The following code snippet creates a PivotTable based on a range of data. The PivotTable has no hierarchies, so the data is not yet grouped in any way.

```typescript
  const dataSheet = workbook.getWorksheet("Data");
  const pivotSheet = workbook.getWorksheet("Pivot");

  const farmPivot = pivotSheet.addPivotTable(
    "Farm Pivot", /* The name of the PivotTable. */
    dataSheet.getUsedRange(), /* The source data range. */
    pivotSheet.getRange("A1") /* The location to put the new PivotTable. */);
```

:::image type="content" source="../images/pivottable-empty.png" alt-text="A PivotTable named 'Farm Pivot' with no hierarchies.":::

### Hierarchies and fields

PivotTables are organized through hierarchies. Those hierarchies are used to pivot data when added as a specific type of hierarchy. There are four types of hierarchies.

- **Row**: Displays items in horizontal rows.
- **Column**: Displays items in vertical columns.
- **Data**: Displays aggregates of values based on the rows and columns.
- **Filter**: Add or removes items from the PivotTable.

> A PivotTable can have as many or as few of its fields assigned to these specific hierarchies. Generally, a useful PivotTable needs at least one data hierarchy to show and at least one row or column to pivot on. The following code snippet adds two row hierarchies and two data hierarchies.

Reviewer 2: It's unclear to me why the PivotTable needs at least one data hierarchy. Could you expand or adjust the phrasing of "a useful PivotTable needs at least one data hierarchy to show"? To *show* what?

:::image type="content" source="../images/pivottable-layout-breakdown.png" alt-text="A diagram showing which sections of a PivotTable are returned by the layout's get range functions.":::

## Filters and slicers

There are three ways to filter a PivotTable.

- [FilterPivotHierarchies](/javascript/api/office-scripts/excelscript/excelscript.filterpivothierarchy)
- [PivotFilters](/javascript/api/office-scripts/excelscript/excelscript.pivotfilters)
- [Slicers](/javascript/api/office-scripts/excelscript/excelscript.slicer)

### FilterPivotHierarchies

`FilterPivotHierarchies` add an additional hierarchy to filter every data row. Any row with an item that is filtered out is excluded from the PivotTable and its summaries. Since these filters are based on items, they only work on discrete values. If "Classification" is a filter hierarchy in our sample, users can select the values of "Organic" and "Conventional" for the filter. Similarly, if "Crates Sold Wholesale" is selected, the filter options would be the individual numbers, such as 120 and 150, instead of numerical ranges.

> FilterPivotHierarchies are created with all values selected. This means that nothing is filtered until the user manually interacts with the filter control or a `PivotManualFilter` is set on the field belonging to the `FilterPivotHierarchy`.

Reviewer 2: ```suggestion
`FilterPivotHierarchies` are created with all values selected. This means that nothing is filtered until the user manually interacts with the filter control or a `PivotManualFilter` is set on the field belonging to the `FilterPivotHierarchy`.
```

> This following code snippet adds "Classification" as a filter hierarchy.

Reviewer 2: ```suggestion
The following code snippet adds "Classification" as a filter hierarchy.
```

```typescript
  farmPivot.addFilterHierarchy(farmPivot.getHierarchy("Classification"));
```

:::image type="content" source="../images/pivottable-filter-hierarchy.png" alt-text="A filter control that uses 'Classification' for a PivotTable.":::

### PivotFilters

The `PivotFilters` object is a collection of filters applied to a single field. Since each hierarchy has exactly one field, you should always use the first field in `PivotHierarchy.getFields()` when applying filters. There are four filter types.

> **Date Filter**: Calendar date-based filtering.

Reviewer 2: These filters are displaying as a paragraph, but I suspect they are intended to display as a list. Also, should both words be capitalized? Suggesting these changes for all 4 filter types. 
```suggestion
- **Date filter**: Calendar date-based filtering.
```

**Label Filter**: Text comparison filtering.
**Manual Filter**: Custom input filtering.
**Value Filter**: Number comparison filtering. This compares items in the associated hierarchy with values in a specified data hierarchy.

> Typically, only one of the four types of filters is created and applied to the field. If the script tries to use incompatible filters, an exception is thrown.

Reviewer 2: Is a specific error thrown?

  classificationField.applyFilter({
    manualFilter: { 
      selectedItems: ["Organic"] /* The included items. */
    }
  });

  const farmField = farmPivot.getHierarchy("Farm").getFields()[0];
  farmField.applyFilter({
    valueFilter: {
      condition: ExcelScript.ValueFilterCondition.greaterThan, /* The relationship of the value to the comparator. */
      comparator: 300, /* The value to which items are compared. */
      value: "Sum of Crates Sold Wholesale" /* The name of the data hierarchy. Note the "Sum of" prefix. */
      }
  });
```

:::image type="content" source="../images/pivottable-filters.png" alt-text="A PivotTable after the value filter and manual filter were applied.":::

### Slicers

[Slicers](https://support.microsoft.com/office/249f966b-a9d5-4b0f-b31a-12651785d29d) filter data in a PivotTable (or standard table). They are moveable objects in the worksheet that allow for quick filtering selections. A slicer operates in a similar fashion to the manual filter and `PivotFilterHierarchy`. Items from the `PivotField` are toggled to include or exclude them from the PivotTable.

> The following code snippet adds a slicer for the "Type" field. It then sets the selected items to be "Lemon" and "Lime", then moves the slicer 400 pixels to the left.

Reviewer 2: ```suggestion
The following code snippet adds a slicer for the "Type" field. It sets the selected items to be "Lemon" and "Lime", then moves the slicer 400 pixels to the left.
```

```typescript
  const fruitSlicer = pivotSheet.addSlicer(
    farmPivot, /* The table or PivotTale to be sliced. */
    farmPivot.getHierarchy("Type").getFields()[0] /* What source to use as the slicer options. */
  );
  fruitSlicer.selectItems(["Lemon", "Lime"]);
  fruitSlicer.setLeft(400);
```

:::image type="content" source="../images/slicer.png" alt-text="A slicer filtering data on a PivotTable.":::

## See also

> - [Scripting Fundamentals for Office Scripts in Excel on the web](scripting-fundamentals.md)

Reviewer 2: ```suggestion
- [Scripting fundamentals for Office Scripts in Excel on the web](scripting-fundamentals.md)
```
